modules/text_server_adv/text_server_adv.cpp

```diff
@@ -1658,6 +1658,123 @@ float TextServerAdvanced::shaped_text_tab_align(RID p_shaped, const Vector<float
 	return 0.f;
 }
 
+void TextServerAdvanced::shaped_text_overrun_trim_to_width(RID p_shaped_line, float p_width, uint8_t p_clip_flags) {
+	_THREAD_SAFE_METHOD_
+	ShapedTextDataAdvanced *sd = shaped_owner.getornull(p_shaped_line);
+	ERR_FAIL_COND_MSG(!sd, "ShapedTextDataAdvanced invalid.");
+	if (!sd->valid) {
+		shaped_text_shape(p_shaped_line);
+	}
+
+	bool add_ellipsis = (p_clip_flags & OVERRUN_ADD_ELLIPSIS) == OVERRUN_ADD_ELLIPSIS;
+	bool cut_per_word = (p_clip_flags & OVERRUN_TRIM_WORD_ONLY) == OVERRUN_TRIM_WORD_ONLY;
+	bool enforce_ellipsis = (p_clip_flags & OVERRUN_ENFORCE_ELLIPSIS) == OVERRUN_ENFORCE_ELLIPSIS;
+
+	Glyph *sd_glyphs = sd->glyphs.ptrw();
+
+	if (sd_glyphs == nullptr || p_width <= 0 || !(sd->width > p_width || enforce_ellipsis)) {
+		return;
+	}
+
+	int sd_size = sd->glyphs.size();
+
+	RID last_gl_font_rid = sd_glyphs[sd_size - 1].font_rid;
+	int last_gl_font_size = sd_glyphs[sd_size - 1].font_size;
+	uint32_t dot_gl_idx = font_get_glyph_index(last_gl_font_rid, '.');
+	Vector2 dot_adv = font_get_glyph_advance(last_gl_font_rid, dot_gl_idx, last_gl_font_size);
+	uint32_t whitespace_gl_idx = font_get_glyph_index(last_gl_font_rid, ' ');
+	Vector2 whitespace_adv = font_get_glyph_advance(last_gl_font_rid, whitespace_gl_idx, last_gl_font_size);
+
+	int ellipsis_advance;
+	if (add_ellipsis) {
+		ellipsis_advance = 3 * dot_adv.x + font_get_spacing_glyph(last_gl_font_rid) + (cut_per_word ? whitespace_adv.x : 0);
+	} else {
+		ellipsis_advance = 0;
+	}
+
+	int ell_min_characters = 6;
+	float width = sd->width;
+
+	int clip_pos = -1;
+	int ellipsis_pos = -1;
+	if (enforce_ellipsis) {
+		clip_pos = 0;
+		ellipsis_pos = 0;
+	}
+
+	int last_valid_cut = 0;
+	bool found = false;
+
+	for (int i = sd_size - 1; i >= 0; --i) {
```

Comment:
This should run from the other end (and add ellipsis on the left side) if detected text direction (`sd->para_direction`) is RTL.

<img width="380" alt="Screenshot 2021-07-02 at 19 23 47" src="https://user-images.githubusercontent.com/7645683/124303672-6062bf80-db6b-11eb-9437-ec3e3634b4b0.png">

Reply:
Implemented.